src/minio/core/policy_builder.py

```diff
@@ -0,0 +1,189 @@
+"""
+Policy Builder module for creating and modifying MinIO policy documents.
+"""
+
+import logging
+import re
+from copy import deepcopy
+from typing import Annotated
+
+from pydantic import BaseModel, ConfigDict, Field
+
+from ...service.exceptions import PolicyOperationError
+from ..models.policy import (
+    PERMISSION_LEVEL_ACTIONS,
+    PolicyAction,
+    PolicyEffect,
+    PolicyModel,
+    PolicyPermissionLevel,
+    PolicyStatement,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class PathAccess(BaseModel):
+    """Represents access to a specific S3 path with permission level."""
+
+    model_config = ConfigDict(frozen=True)
+
+    path: Annotated[str, Field(description="S3 path for access control", min_length=1)]
+    permission_level: Annotated[
+        PolicyPermissionLevel, Field(description="Level of access permission")
+    ]
+
+
+class PolicyBuilder:
+
+    def __init__(self, policy_model: PolicyModel, bucket_name: str):
+        """
+        Initialize the PolicyBuilder with a policy model and bucket name.
+
+        Args:
+            policy_model: The base policy model to modify
```

Comment:
Reading the code, I think the input policy model is modified in place. Maybe rather than deep copying on build, might it makes more sense to deep copy on PolicyBuilder instantiation to avoid the side effect of changes to the input model?

Reply:
👍 updated.

Reply:
:+1: